File: overlay/clojure/hydra-kernel/src/main/clojure/hydra/overlay/clojure/lib/regex.clj
(ns hydra.overlay.clojure.lib.regex
  (:import [java.util.regex Pattern Matcher])
  (:require [hydra.parse.regex :refer :all]
            [hydra.print.pcre.regex :refer :all]))

;; Patterns are Hydra-defined and translingual (docs/specification/regex.md). Each primitive first
;; runs the pattern through hydra.parse.regex, then renders the AST to PCRE syntax via
;; hydra.print.pcre.regex (Clojure regexes are java.util.regex, PCRE-like), before handing the
;; rendered pattern to the native engine. An ill-formed pattern (rejected by hydra.parse.regex) is
;; treated as "no match" -- the same portable-failure convention as an empty match. See issue #603.

;; Returns [:given <native-pattern-string>], or [:none] if the pattern does not parse.
(defn- hydra--regex-to-native [pattern]
  (let [parsed (hydra_parse_regex_parse_regex pattern)]
    (if (= (first parsed) :given)
      [:given (hydra_print_pcre_regex_print_regex (second parsed))]
      [:none])))

;; matches :: String -> String -> Bool
(def hydra_overlay_clojure_lib_regex_matches
  (fn [pattern] (fn [input]
    (let [native (hydra--regex-to-native pattern)]
      (if (= (first native) :none)
        false
        (boolean (re-matches (re-pattern (second native)) input)))))))

;; find :: String -> String -> Maybe String
(def hydra_overlay_clojure_lib_regex_find
  (fn [pattern] (fn [input]
    (let [native (hydra--regex-to-native pattern)]
      (if (= (first native) :none)
        nil
        (let [m (re-find (re-pattern (second native)) input)]
          (if (some? m)
            (if (string? m) m (first m))
            nil)))))))

;; find_all :: String -> String -> [String]
(def hydra_overlay_clojure_lib_regex_find_all
  (fn [pattern] (fn [input]
    (let [native (hydra--regex-to-native pattern)]
      (if (= (first native) :none)
        ()
        (let [results (re-seq (re-pattern (second native)) input)]
          (if results
            (map (fn [m] (if (string? m) m (first m))) results)
            ())))))))

;; replace :: String -> String -> String -> String
(def hydra_overlay_clojure_lib_regex_replace
  (fn [pattern] (fn [replacement] (fn [input]
    (let [native (hydra--regex-to-native pattern)]
      (if (= (first native) :none)
        input
        (let [m (re-matcher (re-pattern (second native)) input)]
          (if (.find m)
            (str (.substring input 0 (.start m))
                 replacement
                 (.substring input (.end m)))
            input))))))))

;; replace_all :: String -> String -> String -> String
(def hydra_overlay_clojure_lib_regex_replace_all
  (fn [pattern] (fn [replacement] (fn [input]
    (let [native (hydra--regex-to-native pattern)]
      (if (= (first native) :none)
        input
        (.replaceAll (re-matcher (re-pattern (second native)) input)
                     (Matcher/quoteReplacement replacement))))))))

;; split :: String -> String -> [String]
(def hydra_overlay_clojure_lib_regex_split
  (fn [pattern] (fn [input]
    (let [native (hydra--regex-to-native pattern)]
      (if (= (first native) :none)
        (list input)
        (seq (.split (Pattern/compile (second native)) input -1)))))))

File: overlay/scala/hydra-kernel/src/main/scala/hydra/overlay/scala/lib/regex.scala
package hydra.overlay.scala.lib

import _root_.java.util.regex.{Matcher, Pattern}

// Patterns are Hydra-defined and translingual (docs/specification/regex.md). Each primitive first
// runs the pattern through hydra.parse.regex, then renders the AST to PCRE syntax via
// hydra.print.pcre.regex (Scala uses java.util.regex, a PCRE-like engine), before handing the
// rendered pattern to the native engine. An ill-formed pattern (rejected by hydra.parse.regex) is
// treated as "no match" — the same portable-failure convention as an empty match. See issue #603.
object regex:
  private def toNative(pattern: String): Option[String] =
    hydra.parse.regex.parseRegex(pattern).map(hydra.print.pcre.regex.printRegex)

  def matches(pattern: String)(input: String): Boolean =
    toNative(pattern) match
      case None => false
      case Some(native) => Pattern.matches(native, input)

  def find(pattern: String)(input: String): Option[String] =
    toNative(pattern) match
      case None => None
      case Some(native) =>
        val m = Pattern.compile(native).matcher(input)
        if m.find() then Some(m.group()) else None

  def findAll(pattern: String)(input: String): Seq[String] =
    toNative(pattern) match
      case None => Seq()
      case Some(native) =>
        val m = Pattern.compile(native).matcher(input)
        val results = _root_.scala.collection.mutable.ArrayBuffer[String]()
        while m.find() do results += m.group()
        results.toSeq

  def replace(pattern: String)(replacement: String)(input: String): String =
    toNative(pattern) match
      case None => input
      case Some(native) =>
        Pattern.compile(native).matcher(input).replaceFirst(Matcher.quoteReplacement(replacement))

  def replaceAll(pattern: String)(replacement: String)(input: String): String =
    toNative(pattern) match
      case None => input
      case Some(native) =>
        Pattern.compile(native).matcher(input).replaceAll(Matcher.quoteReplacement(replacement))

  def split(pattern: String)(input: String): Seq[String] =
    toNative(pattern) match
      case None => Seq(input)
      case Some(native) => Pattern.compile(native).split(input, -1).toSeq
